### PR TITLE
primitive wrapper types for u16 & u32

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
## Description

"new" types to ensure proper conversion of packet data to/from network byte order.

Includes:

- impls bit and/or/xor/*assign and not (!) for types
- includes updates throughout the packets module to use these types
- added flag tests

**Notes**:

- currently includes #94 and will be rebased accordingly after that one is reviewed/merged. 

## Checklist

- [X] Associated tests
- [X] Associated documentation

### CODE REVIEW

File: core/src/packets/types.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+impl ops::Not for u16be {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        u16be(!self.0)
+    }
+}
+
+impl ops::Shl<u16be> for u16be {
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply core/src/packets/types.rs:101]: we don't. I'll remove them.

File: core/src/packets/icmp/v6/ndp/redirect.rs [hunk 2]
```diff
@@ -19,7 +19,7 @@
 use super::{NdpPacket, RedirectedHeader};
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply core/src/packets/icmp/v6/ndp/redirect.rs:19]: think i missed it earlier, but yeah.

File: core/src/packets/types.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply core/src/packets/types.rs:101]: yeah, i was gonna go back to these, cause i thought about using it, but didn't.

File: core/src/packets/types.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+///
+/// Used to convert packet fields to host byte order on get and network byte
+/// order on set.
+#[allow(non_camel_case_types)]
+#[derive(Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(C, packed)]
+pub struct u16be(pub u16);
+
+impl u16be {
+    /// The 16-bit unsigned integer in big-endian order.
```
Comment (drunkirishcoder) [inline core/src/packets/types.rs:36]: I think you made this change in the wrong place. this is for the main type description above. description here should be,

> The smallest value that can be represented by this integer type.

File: core/src/testils/criterion.rs [hunk 4]
```diff
@@ -34,10 +34,10 @@ use std::time::{Duration, Instant};
 pub trait BencherExt {
     /// Times a `routine` with an input generated via a `proptest strategy`
     /// batch of input, and then times the iteration of the benchmark over the
-    /// input. See [`BatchSize`] for details on choosing the batch size. The
+    /// input. See [BatchSize] for details on choosing the batch size. The
```
Comment (drunkirishcoder) [inline core/src/testils/criterion.rs:37]: this is actually a rust type.

File: core/src/testils/proptest/mod.rs [hunk 5]
```diff
@@ -16,12 +16,12 @@
 * SPDX-License-Identifier: Apache-2.0
 */
 
-//! Defines extended [`proptest`] [`Arbitrary`] traits and [`Strategies`] for
+//! Defines extended [proptest] [Arbitrary] traits and [Strategies] for
```
Comment (drunkirishcoder) [inline core/src/testils/proptest/mod.rs:19]: Arbitrary and Strategy are both traits. Maybe change Strategies to singular? Something like 

> Defines extended proptest `Arbitrary` and `Strategy` traits...

File: core/src/packets/types.rs (hunk 3 above)
Comment (zeeshanlakhani) [reply core/src/packets/types.rs:36]: derp.

File: core/src/testils/criterion.rs (hunk 4 above)
Comment (zeeshanlakhani) [reply core/src/testils/criterion.rs:37]: diff lib, but ok, yeah.

**Pull request merged**